Implement layered storage format

hlinnaka opened the pull request:
This evolved from https://github.com/zenithdb/zenith/pull/299, but there has so many changes from the original in-memory repository that I thought a new PR is in order.

This replaces the RocksDB based implementation with an approach using
"snapshot files" on disk, and in-memory btreemaps to hold the recent
changes.
    
This make the repository implementation a configuration option. You can
choose 'layered' or 'rocksdb' in the "pageserver init" call, but there is
no corresponding --repository-formt option in 'zenith init', so in
practice you have to change the default in pageserver.rs if you want to test
different implementations. The unit tests have been refactored to exercise
both implementations, though. 'layered' is now the default.
    
TODOs:
    
- Push/pull is not implemented, causing 'test_history_inmemory' test in 'cargo test' to fail. I think we'll have to rethink how push/pull is implemented. It would make sense to copy the immutable files directly, instead of serializing them into a stream of changes. The test is marked as 'ignore' in this PR.

hlinnaka commented:
With this, we can start working on managing the snapshot files with S3. Uploading the snapshot files to S3, and downloading missing files on demand or wholesale at server startup. Also, I think it's time to start doing performance testing. Performance probably sucks, but we need to start measuring to improve it.

Garbage collection needs to be implemented. I think it's just a matter of removing snapshot files that are old enough. As this stands, the snapshot files always contain full contents of the relation, there are no "incremental" layers. Some logic is needed to not remove files that are still needed by a timeline that was branched off the current one.

The push/pull functionality needs a rethink. I think we want to base that on the snapshot files, too. I think we should comment out the tests related to that for now, and work on fixing them after this is pushed.

hlinnaka commented:
I implemented support for relation drops, and a basic Garbage Collection function, which is roughly on par with the the ObjectRepository's GC implemention. Tests are passing now.

I think this is close to committable state now. I'm doing self-review at the moment, tidying things up. One open question is whether we want to go all in with this approach now, and remove all the RocksDB code as part of this, or keep it as a compile-time option for a little longer, like it is in this PR currently. I'd like to remove it all, but it's also not hard to keep it for a while if it helps with performance testing for example.

Push/pull functionality is not implemented, but I don't think that needs to be a blocker, we can fix or reimplement that as a followup PR.

Please review, everyone!

knizhnik commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,833 @@
+        let snapfiles = std::mem::take(&mut *layers);
+        for snapfile in snapfiles.0.values() {
+            if !snapfile.is_frozen() {
+                snapfile.freeze(last_valid_lsn + 1)?;
```
I wonder why it is needed to add 1 to last_valid_lsn here?

hlinnaka replied:
I've been sloppy on whether the start and end bounds are inclusive or exclusive. This is one such instance. TODO: make sure we're consistent on that, check if this + 1 still makes sense and remove or add a comment.

hlinnaka replied:
Fixed this.

knizhnik commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -0,0 +1,407 @@
+
+    /// If this relation was dropped, remember when that happened. Lsn(0) means
+    /// it hasn't been dropped
+    drop_lsn: Mutex<Lsn>,
```
I wonder why do we use three mutexes here for different fields rather than extract all shared data i one structure and use one mutex for synchronizing access to it.

hlinnaka commented:
I rebased this over PR #351. PR #351 is still under discussion, but this is a step towards merging this into 'main'. With this, the layered storage format handles non-relations, too.

hlinnaka commented:
Rebased this again. Next steps before this is pushed to main:

1. Per offline discussions with people, we don't want to keep maintaining the rocksdb implementation once this lands. So I'm going to remove the configuration option and the code that tries to deal with both implementations, and remove rocksdb as part of this PR.
2. Clean up the commit history, squashing things into one or two commits. 
3. More review, please

knizhnik commented:
> Rebased this again. Next steps before this is pushed to main:
> 
>     1. Per offline discussions with people, we don't want to keep maintaining the rocksdb implementation once this lands. So I'm going to remove the configuration option and the code that tries to deal with both implementations, and remove rocksdb as part of this PR.
> 
>     2. Clean up the commit history, squashing things into one or two commits.
> 
>     3. More review, please

i wonder why we are so hurrying to through away RocksDB implementation?
I agree, that we should replace it, but it will be nice to keep it just as reference implementation for example
for comparing performance.

Right now at buld_insert test layered storage as ~3 times slower than RocksDB, while as far as I understand, unlike RockDB it doesn't provide and consistency guarantees (so in case of page server node abnormal restart, we will have to rebuild all its storage from scratch).

knizhnik commented:
bulk_insert tests results (just insert one million records with sequential keys in table with indexes).
vanilla:    14 seconds
in-mem:  26 seconds
rocksdb:  43 seconds
layred:     84 seconds

knizhnik commented:
According to perf profile a lot of the time is spent in directory traversal. I suspect that it is in `list_snapshot_files` method.
I think that we should keep all metadata in memory and load it only on login. It should no so large (number of segments can not be large) and should fit in memory. And permanent directory traversal seems to be very inefficient.

knizhnik commented:
pgbench scale=1, clients=10, shared_buffers=1Gb (simulating workload which completely fills in compute node's memory)

vanilla: 46205
rocksdb: 41987
layered: 22780

hlinnaka commented:
> i wonder why we are so hurrying to through away RocksDB implementation?

It's a fair amount of extra code we need to maintain and understand, if we keep both.

> I agree, that we should replace it, but it will be nice to keep it just as reference implementation for example
> for comparing performance.

Ok, let's keep the rocksdb implementation for a little longer, for performance testing.

hlinnaka commented:
> According to perf profile a lot of the time is spent in directory traversal. I suspect that it is in `list_snapshot_files` method.
> I think that we should keep all metadata in memory and load it only on login. It should no so large (number of segments can not be large) and should fit in memory. And permanent directory traversal seems to be very inefficient.

Yeah. I noticed that when I saw that it timed out in the CI, and refactored it doing that way.

hlinnaka commented:
> Right now at buld_insert test layered storage as ~2 times slower than RocksDB, while as far as I understand, unlike RockDB it doesn't provide and consistency guarantees (so in case of page server node abnormal restart, we will have to rebuild all its storage from scratch).

The layered storage format is crash safe. The files that are stored on disk are immutable and don't need to be rebuilt on crash. Only the recent WAL that hasn't been written out to files needs to be re-digested. (fsync() calls are currently missing, though, and I haven't done any actual testing of crash-safety, so there probably are bugs)

knizhnik commented:
With in-memory layer list results are much better and comparable with rocksdb

knizhnik commented on pageserver/src/branches.rs:
```diff
@@ -104,19 +104,28 @@ pub fn create_repo(
-        wal_redo_manager,
-        tenantid,
-    );
+    let repo: Arc<dyn Repository + Sync + Send> = match conf.repository_format {
```
There is the same code for creating object repository in repository.rs.
Is there any reason for such code duplication?

knizhnik commented:
I wonder why there is no background GC for layered repository?
Only explicit GC started though page server.

hlinnaka commented:
On 02/08/2021 17:55, Konstantin Knizhnik wrote:
> I wonder why there is no background GC for layered repository?
> Only explicit GC started though page server.

Not implemented yet, that's all.

- Heikki

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+        };
+        LayeredRepository::save_metadata(self.conf, self.timelineid, self.tenantid, &metadata)?;
+
+        // If there were any concurrent updates on the timeline, we would have to work
+        // harder to make sure we don't lose the new updates. Currently, that shouldn't
+        // happen, because the WAL receiver process is responsible for both updating
+        // the timeline and calling checkpoint()
+        assert!(self.last_valid_lsn.load() == last_valid_lsn);
```
So, we don't advance WAL while "checkpoint" is in progress?
To me, it looks like a potential bottleneck.

hlinnaka replied:
Yeah, we surely need to improve that in the future.

That assertion is actually obsolete, the code should handle new updates coming in while checkpoint is in progress now. I'll remove it. But the fact remains that the same thread does both WAL receiving and checkpointing, so it is a potential bottleneck.

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+        let old = self.last_valid_lsn.advance(lsn);
+
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last valid LSN backwards (was {}, new {})",
+                old, lsn
+            );
```
Why does the comment say " Can't move backwards"?

In my understanding, this situation can legally happen if we restart streaming from last_record_lsn and process some record that wasn't fully processed before.

hlinnaka replied:
Ah, you're right, I'll tone it down.

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+    //
+    last_valid_lsn: SeqWait<Lsn>,
+    last_record_lsn: AtomicLsn,
+    prev_record_lsn: AtomicLsn,
```
`prev_record_lsn` deserves a comment too

hlinnaka replied:
Added

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+        rel: RelishTag,
+        lsn: Lsn,
+    ) -> Result<Option<(Arc<dyn Layer>, Lsn)>> {
+        // First dig the right ancestor timeline
```
What does this comment refer to?

hlinnaka replied:
It actually refers to the code a few lines below. Moved things around to make it more clear.

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+        Ok(layer_rc)
+    }
+
+    ///
+    ///
+    ///
```
Huh?

hlinnaka replied:
Added a comment

lubennikovaav reviewed (changes requested):
@hlinnaka, please update the PR description. I guess some of the TODOs are done.


I've looked through most of the files and left some comments. I'm planning to continue the review and testing tomorrow.

With this amount of FIXME, this PR doesn't seem to be ready for merge.
I suggest that we at least create an issue for each FIXME in the code that won't be fixed before merge. It would be good to understand the real amount of work to do.

knizhnik commented:
> On 02/08/2021 17:55, Konstantin Knizhnik wrote: I wonder why there is no background GC for layered repository? Only explicit GC started though page server.
> Not implemented yet, that's all. - Heikki

I have added launching GC thread (branch layered-repo-gc) but it still doesn't work...

hlinnaka commented on pageserver/src/walreceiver.rs:
```diff
@@ -253,7 +254,7 @@ fn walreceiver_main(
                         tenantid,
                     )?;
 
-                    if newest_segno - oldest_segno >= 10 {
+                    if newest_segno - oldest_segno >= 1 {
```
Just to exercise the checkpointing codepath more often, to get more testing on it. I'll revert it.

knizhnik commented:
Good news is that in layered_storage GC has almost now influence on performance, unlike RocksDB.

hlinnaka commented on pageserver/src/layered_repository/README.md:
```diff
@@ -0,0 +1,274 @@
+recent changes to 'customers' table, you would have these files on
+disk:
+
+	main/orders_100_200
+	main/orders_200_300
```
Both, actually. Currently, the code will search for it in the second file, and that's more efficient because then you don't need to replay any WAL to reconstruct the page. But the first file contains all the WAL needed to reconstruct it too. We could optimize and make the first one open. I'll add a note about that to README.

hlinnaka commented on pageserver/src/layered_repository/README.md:
```diff
@@ -0,0 +1,274 @@
+	main/orders_300_400
+
+And there is a branch or explicit recovery point at LSN 150, we could
+replace 'main/orders_100_200' with 'main/orders_150_150' to keep a
```
Correct.

hlinnaka commented:
> > On 02/08/2021 17:55, Konstantin Knizhnik wrote: I wonder why there is no background GC for layered repository? Only explicit GC started though page server.
> > Not implemented yet, that's all. 
> 
> I have added launching GC thread (branch layered-repo-gc). May be you can reuse it.

Ah, thanks, I'll take a look!

knizhnik commented:
Updated performance results:

### Bulk insert: insert 1 million records with sequential keys, shared_buffers=1Mb

**vanilla**: 14 seconds
**in-mem**: 26 seconds
**rocksdb**: 43 seconds
**layered**: 69 seconds

### pgbench: scale=1, clients=10, shared_buffers=1Gb (simulating workload which completely fits in compute node's memory)

**vanilla**: 46205
**rocksdb**: 41987
**layered**: 40063

knizhnik commented:
> > Right now at buld_insert test layered storage as ~2 times slower than RocksDB, while as far as I understand, unlike RockDB it doesn't provide and consistency guarantees (so in case of page server node abnormal restart, we will have to rebuild all its storage from scratch).
> 
> The layered storage format is crash safe. The files that are stored on disk are immutable and don't need to be rebuilt on crash. Only the recent WAL that hasn't been written out to files needs to be re-digested. (fsync() calls are currently missing, though, and I haven't done any actual testing of crash-safety, so there probably are bugs)

Sorry, I do not completely understand it.
As far as I understand looking at layered_repository code, snapshot is saved to the disk as two serialized B-Trees.
What will happen if one of this files will not be written completely?

As far as I see we are not trying to handle this errors. And serde itself doesn't provide any guarantees  that deserialized data is correct. Its documentation says that deserialize **can fail** if input data stream is not correct. But it is not calculating CRC or any other checksum which can detect data corruption.

Also, one of the declared motivations of switching from RocksDB to own storage was that we should control storage format.But in this case we rely on serde serializer and once it is changed, we will not be able to read prevouisly saved data. Certainly in future we can write our own serializer or just fix serde version, but ...

hlinnaka commented:
On 03/08/2021 12:23, Konstantin Knizhnik wrote:
> As far as I understand looking at layered_repository code, snapshot is 
> saved to the disk as two serialized B-Trees.
> What will happen if one of this files will not be written completely?

Ah yeah. So it's not crash safe at the moment :-). But that's easy to 
fix: we need to add CRCs to the files.

> Also, one of the declared motivations of switching from RocksDB to own 
> storage was that we should control storage format.But in this case we 
> rely on serde serializer and once it is changed, we will not be able to 
> read prevouisly saved data. Certainly in future we can write our own 
> serializer or just fix serde version, but ...

The contents of the files was something I could hack together with the 
least effort. A lot of further work is needed there. For starters, the 
file format should allow reading individual pages without slurping the 
whole file into memory. And it should have a header with information 
about the contents, and the CRC. That's all TODOs.

- Heikki

knizhnik commented:
Yeh...
Looks likes layered storage really suffers from write amplification.
`pgbench -i -s 100` takes 1305 seconds with 1Gb shared buffers, comparing with 111 seconds for RocksDB.
Updated test and FameGraph are attached.
But it is not the only bad news: third iteration of read-only pgbench in this test is just crashed. And I still do not find a reasons: there are no core, not errors in pageserver.log and pg.log was overwritten.

[pgbench.tar.gz](https://github.com/zenithdb/zenith/files/6924396/pgbench.tar.gz)

hlinnaka commented:
> Yeh...
> Looks like layered storage really suffers from write amplification.
> `pgbench -i -s 100` takes 1305 seconds with 1Gb shared buffers, comparing with 111 seconds for RocksDB.

Oh yeah, that doesn't surprise me. You can alleviate that somewhat by increasing the "checkpoint interval" in walreceiver_main() from the current hardcoded value of 10 segments. But then it will use more memory.

To be clear: this implementation is just a starting point. We will need a better format then serde-serialized BTreeMaps for the contents. And we will need better strategies for when to dump data to the files. And we will need some kind of an incremental dump format to reduce the write amplification. And compress. And a dozen other tricks we haven't even thought of yet. But the important thing here is that this format is based on immutable files. Once we have that, we can start working on uploading those files to S3, downloading them back to reinitialize a page server (https://github.com/zenithdb/zenith/issues/340). At the same time, we can work on what we store in the files and when and how (https://github.com/zenithdb/zenith/issues/338, https://github.com/zenithdb/zenith/issues/339).

> Updated test and FameGraph are attached.
> But it is not the only bad news: third iteration of read-only pgbench in this test is just crashed. And I still do not find a reasons: there are no core, not errors in pageserver.log and pg.log was overwritten.

Out-of-memory? This implementation slurps the whole file into memory when it needs to access something. That can consume a lot of memory (https://github.com/zenithdb/zenith/issues/337).

knizhnik commented:
> Out-of-memory? This implementation slurps the whole file into memory when it needs to access something. That can consume a lot of memory (#337).
No, it is not out-of-memory. Although pageserver really consumes a lot of memory: it periodical increased  from 3.4 to 4.5Gb and dropped back (I assume when checkpoint happen). 
With disabled GC (please notice that I am using my branch with background GC), very soon disk space (~200Gb) is exhausted.

knizhnik commented:
> To be clear: this implementation is just a starting point.
This is why I was against dropping RocksDB before we will have something stable and comparable with RocksDB in performance.
Right now layered storage is much faster on read-only workload (thanks to in-memory layers) and GC cause less influence on performance.
But looks like random updates cause significant write amplification and more than 10 times slowdown comparing with RocksDB.

knizhnik commented:
Next attempt to perform pgbench -i  -s 1000
failed with the following stack trace:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007fce46d828b1 in __GI_abort () at abort.c:79
#2  0x00005601ec0723b7 in panic_abort::__rust_start_panic::abort () at library/panic_abort/src/lib.rs:46
#3  0x00005601ec072396 in panic_abort::__rust_start_panic () at library/panic_abort/src/lib.rs:41
#4  0x00005601ec060e99 in std::panicking::rust_panic () at library/std/src/panicking.rs:640
#5  0x00005601ec060e54 in std::panicking::rust_panic_with_hook () at library/std/src/panicking.rs:610
#6  0x00005601ec060967 in std::panicking::begin_panic_handler::{{closure}} () at library/std/src/panicking.rs:495
#7  0x00005601ec05e77c in std::sys_common::backtrace::__rust_end_short_backtrace () at library/std/src/sys_common/backtrace.rs:141
#8  0x00005601ec0608f9 in rust_begin_unwind () at library/std/src/panicking.rs:493
#9  0x00005601ebb14031 in core::panicking::panic_fmt () at library/core/src/panicking.rs:92
#10 0x00005601ebb13f7d in core::panicking::panic () at library/core/src/panicking.rs:50
#11 0x00005601ebb776ff in <pageserver::layered_repository::LayeredTimeline as pageserver::repository::Timeline>::checkpoint (self=<optimized out>)
    at pageserver/src/layered_repository.rs:681
#12 0x00005601ebb791d6 in pageserver::layered_repository::LayeredTimeline::gc_timeline (self=0x7fce3401a930, retain_lsns=Vec(size=0), cutoff=...)
    at pageserver/src/layered_repository.rs:1023
#13 0x00005601ebb738de in <pageserver::layered_repository::LayeredRepository as pageserver::repository::Repository>::gc_iteration (self=<optimized out>, target_timelineid=..., 
    horizon=67108864, _compact=<optimized out>) at pageserver/src/layered_repository.rs:231
#14 0x00005601ebb74234 in pageserver::layered_repository::LayeredRepository::gc_loop (self=0x5601ee0ca8b0, conf=<optimized out>) at pageserver/src/layered_repository.rs:257
#15 0x00005601ebb8501a in pageserver::layered_repository::LayeredRepository::launch_gc_thread::{{closure}} () at pageserver/src/layered_repository.rs:249
#16 std::sys_common::backtrace::__rust_begin_short_backtrace (f=...)
    at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125
#17 0x00005601ebbbb2df in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} ()
    at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:481
#18 <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=<optimized out>)
    at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:344
#19 std::panicking::try::do_call (data=<optimized out>) at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:379
#20 std::panicking::try (f=...) at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:343
#21 std::panic::catch_unwind (f=...) at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:431
#22 std::thread::Builder::spawn_unchecked::{{closure}} ()
    at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:480
#23 core::ops::function::FnOnce::call_once{{vtable-shim}} ()
    at /home/knizhnik/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
```

hlinnaka commented:
> #11 0x00005601ebb776ff in <pageserver::layered_repository::LayeredTimeline as pageserver::repository::Timeline>::checkpoint (self=<optimized out>)
>     at pageserver/src/layered_repository.rs:681

That assertion was removed in commit e7c96e935d. Since you're doing the checkpoint from the GC thread, it's not surprising that it triggered. Per the comments there, the checkpointer code assumed that there are new WAL records are not put to the timeline concurrently, and that worked because the WAL receiver thread did the checkpointing. But that assumption wasn't really necessary anymore, I fixed the checkpointing code to work concurrently earlier already, so the assertion could be just removed.

I rebased this again, and added a GC thread similar to you had in your branch. But I also moved the responsibility of checkpointing to the GC thread, that should help to lower the latency of ingesting new WAL records.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+//! same directory that contains information about the timline, in particular its
+//! parent timeline, and the last LSN that has been written to disk.
+//!
+//! This is based on the design at https://github.com/zenithdb/rfcs/pull/8.
```
No, that repo is public. Perhaps we should just remove the link though. The README already covers how this works.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+                let metadata = Self::load_metadata(self.conf, timelineid, self.tenantid)?;
+
+                let ancestor = if let Some(ancestor_timelineid) = metadata.ancestor_timeline {
+                    Some(self.get_timeline_locked(ancestor_timelineid, timelines)?)
```
Hmm, we don't actually lock the ancestor timeline here. `get_timeline_locked` returns an `Arc<LayeredTimeline>`, with no lock. The `_locked` in `get_timeline_locked` means that the caller already holds the lock on the `timelines` map when it calls it. Need a better name for that, I guess.

Perhaps we could avoid looking up the ancestor timeline here, though, and do lazily it if we need to follow the timeline history when looking up a page. Loading the full timeline history into memory could become expensive if you have a very deep history.

hlinnaka replied:
Added comments on that.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+    }
+
+    fn get_page_at_lsn_nowait(&self, rel: RelishTag, blknum: u32, lsn: Lsn) -> Result<Bytes> {
+        if !rel.is_blocky() && blknum != 0 {
```
I'd prefer softer errors, panicing the page server is not nice because it brings it down for all tenants.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+            );
+            result
+        } else {
+            warn!(
+                "get_relsize: rel {} at {}/{} -> not found",
+                rel, self.timelineid, lsn
+            );
+            bail!("relation {} not found at {}", rel, lsn);
```
The warning goes to the log, while bail throws the error up to the caller, all the way to the compute node. Our error handling could be better in general..

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1056 @@
+                    layers.insert(Arc::clone(&new_layer));
+                }
+            } else {
+                // FIXME: if this fails, we're in trouble because we already
+                // swapped the 'layers' with the empty one. Hence panic on error.
+                layer.unload().expect("could not unload layer from memory");
```
It's actually harmless if the 'unload' fails, this could be just a warning to the log instead. However returning with `?` would be bad, because we had already replaced 'layers' with an empty map. It's OK to 'take' the layer map first, because we're holding the mutex on it, but if we do that, we must not return from the function until we're replaced it with `new_layers`. Now that I look at this again, I see that the `layer.freeze(..)?` has the same problem; if that fails, we are left with an empty layers map.

I'll try to refactor that to make it more safe.

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+    /// Launch the checkpointer thread in given repository.
+    ///
+    pub fn launch_checkpointer_thread(conf: &'static PageServerConf, rc: Arc<LayeredRepository>) {
+        let _gc_thread = std::thread::Builder::new()
```
I guess `gc_thread` here is some copy-paste leftover, right?

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+    //   don't need to keep that in the parent anymore. But currently
+    //   we do.
+    //
+    // - Currently, this is only triggered manually by the 'do_gc' command.
+    //   There is no background thread to do it automatically.
```
Outdated comment. GC is called from `checkpointer_loop()`

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+            }
+
+            // Garbage collect old files that are not needed for PITR anymore
+            if conf.gc_horizon > 0 {
```
Should we check for `gc_horizon` somewhere earlier? Maybe don't even start the thread if it is 0.

hlinnaka replied:
We still want to run the checkpointing even if we don't remove old layers.

lubennikovaav commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+
+        info!("saving metadata {}", path.display());
+
+        file.write_all(&TimelineMetadata::ser(data)?)?;
```
Don't we need to call  `file.flush()?;` ?

hlinnaka replied:
Opened https://github.com/zenithdb/zenith/issues/427 to track this

lubennikovaav commented on pageserver/src/walreceiver.rs:
```diff
@@ -264,7 +264,9 @@ fn walreceiver_main(
 
                     if newest_segno - oldest_segno >= 10 {
-                        timeline.checkpoint()?;
+                        // FIXME: The layered repository performs checkpointing in a separate thread, so this
+                        // isn't needed anymore. Remove 'checkpoint' from the Timeline trait altogether? OR
+                        //timeline.checkpoint()?;
```
Since we decided to keep rocksdb for a while, let's keep this with repository format check.
Something like `if repository_format == RepositoryFormat::RocksDb { timeline.checkpoint()?; }`

hlinnaka replied:
done

lubennikovaav commented on pageserver/src/object_repository.rs:
```diff
@@ -691,137 +711,6 @@ impl Timeline for ObjectTimeline {
         Ok(Box::new(ObjectHistory { lsn, iter }))
     }
-
-    fn gc_iteration(&self, horizon: u64, compact: bool) -> Result<GcResult> {
```
Can you split this GC refactoring to a separate smaller PR?

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+            if let Some(fname) = direntry.file_name().to_str() {
+                if let Ok(timelineid) = fname.parse::<ZTimelineId>() {
+                    // Read the metadata of this timeline to get its parent timeline.
+                    // FIXME: we open the timeline below with get_timeline() anyway,
+                    // so why not fetch the metadata from the in-memory Timeline
+                    // struct?
```
Once we have determined that a timeline exists on disk, by scanning the directory, we could use `get_timeline()` to load and access its `LayeredTimeline` struct. I'll try to clarify the comment.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -0,0 +1,1145 @@
+        let now = Instant::now();
+
+        // grab mutex to prevent new timelines from being created here.
+        // TODO: We will hold it for a long time
+        let mut timelines = self.timelines.lock().unwrap();
```
Perhaps. I'm not sure if there would be race conditions if a new branch was created.

One quick thing we could do is to use an RwLock instead of Mutex. That would already help a lot, because then the GC would only block creating new branches but existing branches could be accessed concurrently.

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -0,0 +1,519 @@
+        Ok(())
+    }
+
+    ///
+    /// Write the this in-memory layer to disk, as a snapshot layer.
+    ///
+    /// The cutoff point for the layer that's written to disk is 'end_lsn'.
+    ///
+    /// Returns new layers that replace this one. Always returns a
+    /// SnapshotLayer containing the page versions that were written to disk,
+    /// but if there were page versions newer than 'end_lsn', also return a new
+    /// in-memory layer containing those page versions. The caller replaces
+    /// this layer with the returned layers in the layer map.
+    ///
+    fn freeze(
+        &self,
+        end_lsn: Lsn,
+        walredo_mgr: &dyn WalRedoManager,
+    ) -> Result<Vec<Arc<dyn Layer>>> {
```
Hmm. What is the "latest LSN"? The individual layers don't track such a thing, the closest we have is the last_valid_lsn in LayeredTimeline. And that's what the caller passes here as the end_lsn argument.

One problem is that one WAL record can modify many pages. All the modifications are parsed in `save_decoded_record', which turns one WAL record into a sequence of `put` calls in the Repository. For example, a page split record could result in a call sequence like this:
```
  timeline.advance_last_record_lsn(90) // previous record

  timeline.put_wal_record(rel, 10 /* blk */, 100 /* LSN */);  // left page
  timeline.put_wal_record(rel, 11 /* blk */, 100 /* LSN */);  // right page
  timeline.advance_last_record_lsn(100)
```
The `advance_last_record` calls work like atomic boundaries. If a checkpoint happens in between the two put_wal_record() calls and we need to freeze the layer, we would freeze and create the on-disk layer as of LSN 90 and keep the changes for LSN 100 in the new in-memory layer.

ericseppanen replied:
I'm not sure if I understand.

I think that layer mutation and snapshotting should be atomic relative to one another. I would much prefer a design that will only perform a snapshot when the layers are in a consistent state.

Then the snapshot/write-file process goes like this:

start:

    LAYER3(memory) <- all writes go here.
    LAYER2(file)
    LAYER1(file)

step 2: atomic relative to wal record processing, so no "torn" writes are possible.

    LAYER4(memory) <- all writes now land here
    LAYER3(memory) <- now read-only; this is a temporary state until step 3 finishes.
    LAYER2(file)
    LAYER1(file)

step 3: write out the previous top layer to a file; atomically swap the file `LAYER3'` for the in-memory `LAYER3`.

    LAYER4(memory) <- all writes now land here
    LAYER3'(file)
    LAYER2(file)
    LAYER1(file)

The nice properties of this are:
- Creating a snapshot file doesn't involve splitting the layer into "new" and "old" layers, it's just writing out 100% of the data in that layer.
- Creating a snapshot file doesn't need to copy any `BtreeMap` data structures. Because the map is now read-only, we can just iterate over its contents, streaming its data into a file (with very little memory overhead).
- All of the steps are either fast/atomic (create new top layer) or don't modify the in-memory layer (writing the file), so there is no concurrency with writers. Writers only modify the top layer, while snapshots only deal with layers that are static & unchanging. This makes it a lot easier to ensure correctness because there is minimal opportunity for race conditions.
- The previous layers and the new layers have a 1:1 relationship, which makes debugging simple.
- `LAYER3` and `LAYER3'` contain the same data, so it's really simple to demonstrate that the code is correct.

hlinnaka replied:
+1 on that approach.

A simple way to make step 2 atomic with WAL processing would be to have a global mutex that you hold when you process change for one WAL records, and also in step 2 above. A global mutex seems pretty heavy-weight, but if it's only those two threads that compete for it, and "step 2" is pretty fast, I guess it would be ok.

I'll give that a try.

hlinnaka commented on pageserver/src/layered_repository/storage_layer.rs:
```diff
@@ -0,0 +1,113 @@
+    fn freeze(&self, end_lsn: Lsn, walredo_mgr: &dyn WalRedoManager)
+        -> Result<Vec<Arc<dyn Layer>>>;
+
+    /// Permanently delete this layer
+    fn delete(&self) -> Result<()>;
```
The GC code can figure out when a layer has been superseded by a newer layer and can be deleted. See `LayeredTimeline.gc_timeline`. It calls this.

hlinnaka commented on pageserver/src/layered_repository/storage_layer.rs:
```diff
@@ -0,0 +1,113 @@
+    /// Permanently delete this layer
+    fn delete(&self) -> Result<()>;
+
+    /// Try to release memory used by this layer. This is currently
+    /// only used by snapshot layers, to free the copy of the file
+    /// from memory. (TODO: a smarter, more granular caching scheme
+    /// would be nice)
+    fn unload(&self) -> Result<()>;
```
Yeah, this is just temporary. Currently, the code slurp the whole file into memory. As we get smarter about that, and only keep a little bit of metadata in memory for each file, we probably don't need to "unload" it.

hlinnaka commented on pageserver/src/layered_repository/storage_layer.rs:
```diff
@@ -0,0 +1,113 @@
+    /// records, whereas an unfrozen layer can still be modified, but is not
+    /// durable in case of a crash. Snapshot layers are always frozen, and
+    /// in-memory layers are always unfrozen.
+    fn is_frozen(&self) -> bool;
```
Right. Only one un-frozen top layer for each relish. This returns constant 'true' for InMemoryLayers and constant 'false' for SnapshotLayers. The checkpoint code deals with `dyn Layer`s but needs to somehow distinct between the two, that's what this function is for.

I think it would make sense to keep the InMemoryLayers and SnapshotLayers more separate in the LayerMap. Only the code that wants to read a page should treat them the same, most other code, like opening a layer for writing and checkpointing/freezing the layers, needs to distinguish them. I'll look into refactoring that, and I think this function can then go away too.

hlinnaka commented on pageserver/src/layered_repository/storage_layer.rs:
```diff
@@ -0,0 +1,113 @@
+        lsn: Lsn,
+    ) -> Result<Bytes>;
+
+    fn get_relish_size(&self, lsn: Lsn) -> Result<Option<u32>>;
+
+    fn get_rel_exists(&self, lsn: Lsn) -> Result<bool>;
+
+    fn put_page_version(&self, blknum: u32, lsn: Lsn, pv: PageVersion) -> Result<()>;
+
+    fn put_truncation(&self, lsn: Lsn, relsize: u32) -> anyhow::Result<()>;
+
+    fn put_unlink(&self, lsn: Lsn) -> anyhow::Result<()>;
```
Well, they are two different classes of operations, and I like to group them together somehow. Maybe just comments would do. Or have separate ReadLayer and WriteLayer traits.

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -0,0 +1,519 @@
+pub struct InMemoryLayerInner {
+    /// If this relation was dropped, remember when that happened. Lsn(0) means
+    /// it hasn't been dropped
+    drop_lsn: Lsn,
```
done

ericseppanen replied:
I'll abuse this conversation to mention that it's quite common in Rust to do one-off `Option`-like types, e.g.

```
enum DropStatus {
    Alive,
    Dropped(Lsn),
}
```

Which might be a good idea if you find e.g. `drop_lsn.is_none()` sprinkled all over the place, and are constantly wondering whether `None` means "dropped", or "not-dropped".

I'm not suggesting we do this here, just mentioning it as a general strategy to be aware of...

ericseppanen commented on pageserver/src/layered_repository/storage_layer.rs:
```diff
@@ -0,0 +1,113 @@
+
+pub static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; 8192]);
+
+///
+/// Represents a version of a page at a specific LSN. The LSN is the key of the
+/// entry in the 'page_versions' hash, it is not duplicated here.
+///
+/// A page version can be stored as a full page image, or as WAL record that needs
+/// to be applied over the previous page version to reconstruct this version.
+///
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PageVersion {
+    /// an 8kb page image
+    pub page_image: Option<Bytes>,
+    /// WAL record to get from previous page version to this one.
+    pub record: Option<WALRecord>,
+}
```
This must always be one image OR one record, yes? If so, I think this should be changed to:
```
enum PageVersion {
    Image(Bytes),
    Record(WALRecord),
}
```
This better expresses what we want to allow, because it's impossible to have `{`None, None`}` or `{ Some(...), Some(...)`.

hlinnaka replied:
> This must always be one image OR one record, yes?

Surprisingly, no. When the WAL is received and processed and the wal decoder calls put_*() functions, you only have an image or a record on each PageVersion. However, when a get_page_at_lsn() request comes in and we have to perform WAL redo to reconstruct a page image, we also store the reconstructed image back in the PageVersion. The rocksdb implementation had reasons to do that because of the way the GC is implemented there, see comment in ObjectTimeLine.get_page_at_lsn`. The Layered implementation does the same, although it wouldn't have to.

That's a bit strange when the in-memory layer gets frozen. We will store both the image and the WAL record in the on-disk layer too, if we happened to both in memory.

This is not very well-thought out, we need a more coherent plan on what to materialize and when and which versions to keep as images and which as WAL records. But as things stand, PageVersion can hold both. I'll add a comment on that for now..

hlinnaka commented:
Merged into main. Thanks everyone for the reviews and comments!

hlinnaka commented:
I have opened github issues for a few things that y'all pointed out in the review comments that I didn't address yet.

Next thing that I'm going to work on:
- Divide the key-space into smaller 10 MB segments. (Currently, this treats each whole relish as one big segment)
